[ADR-16] Refactor Folders

Hinton opened the pull request:
## Type of change

<!-- (mark with an `X`) -->

```
- [ ] Bug fix
- [ ] New feature development
- [x] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

<!--Describe what the purpose of this PR is. For example: what bug you're fixing or what new feature you're adding-->

Refactors the Folder domain to follow the new method described in [ADR-16](https://contributing.bitwarden.com/architecture/adr/decryption-in-views). Please read the ADR for some background and the motivation for this move.

This PR focuses on introducing the core services needed for the new data model, and migrates the Folder domain.

The way this is implemented is by adding a generic method `decryptDomain` to `CryptoService`. Which currently fetches the relevant encryption key and calls `decryptDomain` on `EncryptionService`. Which in turns calls the static method `FolderView.decrypt(encryptionService, key, domain)`. Which performs the actual mapping from domain to view.

Fetching of the encryption key is implemented for organizations and personal owned items. Personal items gets a null value assigned while organization items provides the organization ID as the identifier. In the future we can support more keys by adding more lookup locations.

Encryption works in a similar fashion using the `encryptView` which fetches the key and proxies the request to `EncryptionService` which calls `decrypt` on the view.

Why not just put the decrypt on `FolderService`? The `FolderService` already has a responsibility, and that is to maintain the state of `Folders`. We shouldn't give it more responsibilities, particularly not when the `EncryptService` is already responsible for encrypting and decrypting things.

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

eliykat reviewed (commented):
I've also been noodling away on this, please review #3738 as an alternative proposal and let me know what you think.

The biggest difference is that I don't want to have this pattern where we:
* pass an object to a decrypt method
* that method passes a service into a method on the object
* there are several levels of nested decrypt calls, passing down the same parameters each time, which you have to trace to find the actual decryption call still nested in `EncString`.

To me that is the real source of the tech debt here, it feels like decryption happens from the inside out, it's difficult to understand and extend. `ContainerService` is just a symptom of that. I'd like to flatten it out to a method which acts on an object.

Hinton commented on libs/common/src/models/view/folderView.ts:
```diff
@@ -1,11 +1,26 @@
+};
+
+export abstract class DecryptableView extends View {
+  static decrypt: <T extends DecryptableView, D extends Domain>(
+    cryptoService: CryptoService,
+    key: SymmetricCryptoKey,
+    model: D
+  ) => Promise<T>;
```
Yes, we still have type safety though since the method in crypto service enforces the method definition.

Hinton commented:
One goal that hasn't been noted in this PR, is the potential use case for a domain model to have multiple views with different amount of data.

I kept the `decrypt` method for a few reasons,
1. to limit the scope of the refactor.
2. Support objects that decrypt differently attachments seems to have some custom logic, and there has been some discussion around potentially having an EncObject which contains a json blob instead of individual fields.

eliykat commented:
> Support objects that decrypt differently attachments seems to have some custom logic, and there has been some discussion around potentially having an EncObject which contains a json blob instead of individual fields.

Just a note on this topic, I'd like to avoid having lots of custom logic here or different ways of encrypting objects. I'd rather define a few ways of encrypting objects (e.g. individual EncStrings or an encrypted blob) and make people implement those, rather than give every object the ability to do something different.

> One goal that hasn't been noted in this PR, is the potential use case for a domain model to have multiple views with different amount of data.

Nice idea!

MGibson1 commented on libs/common/src/services/crypto.service.ts:
```diff
@@ -678,6 +680,19 @@ export class CryptoService implements CryptoServiceAbstraction {
+    view: DecryptableDecryptType<T, D>,
+    model: D
+  ): Promise<T> {
+    // If the item has an organizationId, use the org key, otherwise use the user key
```
IMO it's the responsibility of the domain to know if it's encrypted and what key will decrypt it.

eliykat replied:
Just to round off this train of thought, if the domain service gets and supplies its own key, then we don't need this method at all. The domain service just calls the static decrypt method on the View object directly.

eliykat replied:
Resolving this as we agreed that this method will live on `CryptoService` for now.

Ideally we'll refactor it later so that encrypt/decrypt can get its own key from `KeyService`, and we have a static `KeyService` implementation that can be pre-seeded with keys for stateless contexts.

MGibson1 commented:
I don't particularly like that it's the `View`'s responsibility to determine what data on the `Domain` is encrypted and what is not. To me, #3738 and this are complimentary -- #3732 describes the data flow and #3738 is a better interface for defining encrypted data/decryption. @eliykat, the big change that would be necessary in your PR is updating the `decryptable` interface to belong on the `View` instead of the `Domain`. This would enable @Hinton's idea of multiple views for a given domain. The issue is how do we expose to the `encryptservice` which fields to decrypt and which are not needed? Not sure on that one, yet...

The other issue I see here is this doesn't solve the central issue of how to grab the key to my liking. I think I'd like to see an interface on `Domain` that determines the appropriate key, given a 2d array of `[Id, Key][]`.

Hinton commented:
I don't mind having an interface on Domain to determine which key is used to encrypt things.

Something that came to mind are Attachments, Name is an `EncString`, but the key has custom logic, and the data is encrypted using the key. Not sure if we have more edge cases for how things are encrypted/decrypted.

eliykat commented on libs/common/src/models/view/folderView.ts:
```diff
@@ -19,6 +34,12 @@ export class FolderView implements View, ITreeNodeObject {
 
+  static async decrypt(cryptoService: CryptoService, key: SymmetricCryptoKey, model: Folder) {
+    const view = new FolderView(model);
+    view.name = await cryptoService.decryptToUtf8(model.name, key);
```
I like pulling the `cryptoService` call up here because we're not descending into the EncString. But we lose the error handling code which catches any error and substitutes `[error: cannot decrypt]`, which imo is important to avoid aborting an entire vault decryption due to 1 corrupted cipher.

Hinton replied:
That seems reasonable, and I agree that does seem kind of nice.

eliykat replied:
I realised that this is arguably putting the decrypt method on the Domain, but given that the "view" is just a `string`, this might just be an exception to the rule. Same with `EncArrayBuffer` -> `ArrayBuffer`. Or maybe the view/domain distinction just doesn't really make sense at this level, so it doesn't matter anyway.

Hinton replied:
Implemented, let me know what you think.

eliykat commented on libs/angular/src/components/folder-add-edit.component.ts:
```diff
@@ -90,7 +92,7 @@ export class FolderAddEditComponent implements OnInit {
       this.editMode = true;
       this.title = this.i18nService.t("editFolder");
       const folder = await this.folderService.get(this.folderId);
-      this.folder = await folder.decrypt();
+      this.folder = await this.cryptoService.decrypt(FolderView, folder);
```
To your point about where decryption happens: I was thinking that `FolderService` should only be handing out View objects, not Domain objects. But then I suppose that leads to a proliferation of methods for different folder view classes. Maybe `this.folderService.get(this.folderId, FolderViewClass)`?

eliykat replied:
I realised this doesn't make any sense with reactive data

eliykat commented on libs/common/src/services/crypto.service.ts:
```diff
@@ -678,6 +680,19 @@ export class CryptoService implements CryptoServiceAbstraction {
+    model: D
+  ): Promise<T> {
+    // If the item has an organizationId, use the org key, otherwise use the user key
+    const orgId: string = (model as any).organizationId;
+    const key = Utils.isNullOrWhitespace(orgId)
+      ? await this.getKeyForUserEncryption()
+      : await this.getOrgKey(orgId);
```
I liked the idea of the `keyId` we discussed, e.g. `const key = this.getKey(model.keyId)`

Where `model` implements an interface for `keyId`, it returns a guid or `null` for user's key.

Hinton replied:
Something that occurred to me is the need for views to know which key to use.

eliykat replied:
For decrypting - view factories get the `keyId` from domain model and passes it to `cryptoService` (or whatever service we end up using) when decrypting 

For encrypting - I suppose the view also implements the same `keyId` interface to persist this information and passes it to the service when encrypting (in a similar way)

(some exceptions apply, e.g. Send/Attachments)

Is this what you mean?

Hinton commented on libs/common/src/services/crypto.service.ts:
```diff
@@ -678,6 +684,26 @@ export class CryptoService implements CryptoServiceAbstraction {
     return true;
   }
 
+  async decrypt<V, D extends DecryptableDomain>(view: Decryptable<V, D>, domain: D): Promise<V> {
+    const key = await this.getKeyFromIdentifier(domain.keyIdentifier());
+
+    return view.decrypt(this, key, domain);
+  }
```
Assuming we want to move this to encryption service, I envision we would lift this block of code, and keep `getKeyFromIdentifier` in the crypto service.

eliykat commented on libs/common/src/models/view/folder.view.ts:
```diff
@@ -19,6 +21,25 @@ export class FolderView implements View, ITreeNodeObject {
+  }
+
+  static async decrypt(cryptoService: CryptoService, key: SymmetricCryptoKey, model: Folder) {
+    const view = new FolderView(model);
```
Are we still using constructors in this pattern? I don't like that it exposes a public method which does not properly initialize the object (because it doesn't handle encStrings). I'd rather move the current constructor logic into this static factory method.

Hinton replied:
That's a fair point, I don't mind removing the constructor.

eliykat reviewed (commented):
Only one minor comment so far, unfortunately I didn't get to spend a lot of time on this today. I'll continue Monday.

eliykat commented on libs/common/src/abstractions/crypto.service.ts:
```diff
@@ -84,4 +90,9 @@ export abstract class CryptoService {
   randomNumber: (min: number, max: number) => Promise<number>;
   validateKey: (key: SymmetricCryptoKey) => Promise<boolean>;
+
+  decrypt: <V, D extends DecryptableDomain>(view: Decryptable<V, D>, model: D) => Promise<V>;
+  encryptView: <V extends Encryptable<EncryptableDomain<V>>>(
```
I wish we had method overloading 😞 I'd like some way of distinguishing this from the existing encrypt/decrypt methods, maybe `decryptItem` and `encryptItem`? (but it's also used for some things like Folders, which are not vault items - any ideas?)

Hinton replied:
`decryptView`  and `encryptView`? I partially feel like we need a low level service that handles primities, and a higher level that handles domain <-> view. Or somehow fit EncString and EncArrayBuffer into the same interface. Perhaps a task for a late date though.

eliykat commented on libs/common/src/models/domain/folder.ts:
```diff
@@ -1,43 +1,27 @@
       return;
     }
 
-    this.buildDomainModel(
```
I also prefer just manually mapping it, it's much more discoverable and straightforward, and gets rid of these opaque mapping methods in the base class.

MGibson1 replied:
I agree

eliykat commented on libs/common/src/models/domain/folder.ts:
```diff
@@ -1,43 +1,27 @@
-    );
-
+    this.id = obj.id;
+    this.name = new EncString(obj.name);
```
`buildDomainModel` has a null check before instantiating the EncString, which we need to put back in. But it's verbose to repeat it on every line, and we can't move it into the constructor (which always creates the object). Maybe a helper method, like `newIfNotNull(val: any, ctor: Constructor)`, so we can use it for dates as well?

Hinton replied:
I managed to add a type safe nullable factory.

eliykat commented on libs/common/src/models/view/encryptable.ts:
```diff
@@ -0,0 +1,42 @@
+  /**
+   * Unique GUID for the key used to encrypt the data
+   */
+  keyIdentifier(): string | null;
```
This would be tidier as a getter. It should never need an argument.

Hinton replied:
So I went back and forth on this a bit. It seems the `fromJson` interface doesn't really like readonly getters. We could probably tweak the type to ignore them but that also feels like a hassle, and it might just be simpler to keep the function. That way we are certain it won't leak into json serializations should someone incorrectly use a property instead of getter.

eliykat commented on libs/common/src/models/view/folder.view.ts:
```diff
@@ -1,22 +1,37 @@
+  async encrypt(cryptoService: CryptoService, key: SymmetricCryptoKey): Promise<Folder> {
+    const folder = new Folder();
+    folder.id = this.id;
+    folder.name = await cryptoService.encrypt(this.name, key);
```
I think it would be clearest if we assigned encrypted & unencrypted properties in separate blocks of code. (i.e. just separated by some whitespace)

eliykat commented on libs/common/src/models/view/folder.view.ts:
```diff
@@ -1,22 +1,37 @@
+  static async decrypt(cryptoService: CryptoService, key: SymmetricCryptoKey, model: Folder) {
+    const view = new FolderView();
+    view.id = model.id;
+    view.name = await model.name.decryptWithCryptoService(cryptoService, key);
```
Need a null check on any EncString method calls

Hinton replied:
We really need to use strict 😢

eliykat commented:
I'm still a little unsure what the end state of `CryptoService` and `EncryptService` is. The duplicated interfaces between them are starting to get messy.

iirc, the last we discussed was potentially changing `CryptoService` to be `KeyService`, then have `EncryptService` depend on `KeyService` so that it can fetch its own keys (with a static key store for web workers). That brought up circular dependency issues that we did not fully resolve. (see my [notes](https://bitwarden.atlassian.net/wiki/spaces/~805680025/pages/283148305/Notes+Item+Decryption#CryptoService%2FKeyService-%E2%86%94%EF%B8%8E-EncryptSevice-dependency)). While we don't need to do all of this in 1 PR, I'd like to know where we're going with it.

Hinton commented:
@eliykat I understand and I didn't want to muddle this PR with any refactoring of `CryptoService`.

I agree with the move to `KeyService` and have `EncryptService` depend on `KeyService`. But I view that as something we'll get around to at a later stage. I would like to refactor `CryptoService` to observables and do this refactor prior which should make that change fairly simple.

Hinton commented on libs/common/src/interfaces/decryptable.interface.ts:
```diff
@@ -7,6 +7,7 @@ import { InitializerMetadata } from "./initializer-metadata.interface";
  * @example Cipher implements Decryptable<CipherView>
  */
-export interface Decryptable<TDecrypted extends InitializerMetadata> extends InitializerMetadata {
+export interface OldDecryptable<TDecrypted extends InitializerMetadata>
```
I didn't have time to figure out a good name for this. But I envision that this needs to be rethought as there won't be a `decrypt` operation on the domain, and we instead want to convey

```json
{
  "domain": cipher,
  "view": CiperView,
}
```

We might need a mapper that converts strings to instances ...

MGibson1 replied:
That is indeed what `InitializeMetadata` does.

I think we need to figure out a way to combine this interface definition with the crypto.interface ones.  The idea was that we eventually want to be able to support any domain/view being encrypted/decrypted in a web worker. 

The big impact is that we need to have (de)serialization guaranteed for any domain/view pair.

I've played a little bit and they're almost compatible. The big additional data is the to/from json information and inclusion of an enum flag to map the appropriate deserializer. There are some issues with `DecryptableDomain` and `Decryptable` separation, but I think those are surmountable.

eliykat replied:
I agree with @MGibson1 above. To expand slightly, I think you'd have `DecryptableDomain` and `Encryptable` extend `InitializerMetadata`, as they represent the instantiated objects that are being passed to/from the web worker. That effectively replaces this `OldDecryptable` interface.

In addition, you'd have an additional `InitializerKey` property on the `DecryptableDomain` to represent the view it decrypts to.

You've reverted the cipher changes, so we can discuss further in the cipher PR.

Hinton replied:
We have a working version of this in https://github.com/bitwarden/clients/pull/4473,

MGibson1 commented on libs/common/src/models/domain/folder.ts:
```diff
@@ -1,47 +1,31 @@
-      null
-    );
+  keyIdentifier(): string | null {
+    return null;
```
please comment that folders are always in the user context

MGibson1 commented on libs/common/src/interfaces/crypto.interface.ts:
```diff
@@ -0,0 +1,53 @@
+/**
+ * View model that encrypts to Domain.
+ */
+export interface Encryptable<Domain> {
```
Please prefix type parameters with `T`

```suggestion
export interface Encryptable<TDomain> {
```

MGibson1 commented on libs/common/src/models/view/folder.view.ts:
```diff
@@ -1,22 +1,38 @@
-      return;
-    }
+  keyIdentifier(): string | null {
+    return null;
```
Please comment that folders are always user-encrypted

Hinton commented on libs/angular/src/components/folder-add-edit.component.ts:
```diff
@@ -42,7 +44,7 @@ export class FolderAddEditComponent implements OnInit {
 
     try {
-      const folder = await this.folderService.encrypt(this.folder);
+      const folder = await this.cryptoService.encryptView(this.folder);
```
I agree. I would probably argue that the save folder action should accept a folder view instead, and delegate the encryption action to the service. But not have an explicit `encyrpt/decrypt` action on the service.

MGibson1 replied:
❤️ sounds much better.

MGibson1 commented on patch.diff:
```diff
@@ -0,0 +1,565 @@
+diff --git a/apps/desktop/desktop_native/Cargo.lock b/apps/desktop/desktop_native/Cargo.lock
```
Looks like you have a patch mistakenly checked in

Hinton commented:
I've removed the `Cipher` changes from this again, and changed the target to a new feature branch. My thinking is that we'll review this to ratify the general patterns, get QA to begin testing folders while I move the cipher portions to a new PR and we can review it prior to merging the feature branch.

This lets QA test both potions in isolation and we don't introduce the change until we've migrated Ciphers which prevents us from having 3 ways of doing encryption/decryption of domain models.

eliykat commented on libs/common/src/models/view/folder.view.ts:
```diff
@@ -1,26 +1,43 @@
+    folder.id = this.id;
+    folder.revisionDate = this.revisionDate;
+
+    folder.name = this.name != null ? await encryptService.encrypt(this.name, key) : null;
```
I think we can also use `NullableFactory` here as well, right?

Hinton replied:
Not really, it's not a constructor. We could make a nullable factory that accepts an argument with a factory in it.

eliykat replied:
True. It's a similar pattern though (call a method if value isn't null), and this ternary condition will be repeated 100 times as we encrypt properties on objects under this new pattern.

eliykat commented on libs/common/src/services/crypto.service.ts:
```diff
@@ -678,6 +684,29 @@ export class CryptoService implements CryptoServiceAbstraction {
+  ): Promise<V> {
+    const key = await this.getKeyFromIdentifier(domain.keyIdentifier());
+
+    return this.encryptService.decryptDomain(view, domain as DecryptableDomain, key);
```
Also, while it's strictly unnecessary, I think we have a practice of always `await`ing async methods, even if it's a tail call. I don't feel strongly about it but it does feel a bit safer to me.

eliykat commented on libs/common/src/interfaces/crypto.interface.ts:
```diff
@@ -0,0 +1,53 @@
+    return null;
+  }
+
+  return new c(...(args as any[]));
```
You shouldn't have to type cast here if you type it like this in the parameters. e.g.

```ts
export function nullableFactory<T extends new (...args: any[]) => any>(
  // <snip>
  return new c(...args);
}
```

eliykat commented on libs/common/src/interfaces/crypto.interface.ts:
```diff
@@ -0,0 +1,53 @@
+export function nullableFactory<T extends new (...args: any) => any>(
+  c: T,
+  ...args: ConstructorParameters<T>
+): InstanceType<T> | undefined {
```
Why `| undefined`? Shouldn't it be `| null`?

eliykat commented on libs/common/src/interfaces/crypto.interface.ts:
```diff
@@ -0,0 +1,53 @@
+import { EncryptService } from "../abstractions/encrypt.service";
+import { SymmetricCryptoKey } from "../models/domain/symmetric-crypto-key";
+
+export function nullableFactory<T extends new (...args: any) => any>(
```
`nullableFactory` seems broadly useful to me, I think it should stand outside of the crypto-specific code.

eliykat reviewed (changes requested):
Mostly minor comments given that the cipher changes have been taken out and I think we've discussed the folder stuff quite a bit already.

To be honest I'm struggling with the various type/interface names (`Decryptable`/`Encryptable`) etc, but I don't have any better suggestions for them. I'll chew on it though.

Hinton commented on libs/common/src/interfaces/crypto.interface.ts:
```diff
@@ -0,0 +1,53 @@
+  /**
+   * Unique GUID for the key used to encrypt the data
+   */
+  keyIdentifier(): string | null;
```
As previously covered, making it a getter will include it in serialization. A method lives on the prototype which avoids that hassle.

eliykat replied:
ah right, yes I thought I mentioned it before but didn't see your response.

Hinton commented:
We can discuss it in the future PR, but I don't want to mix in any knowledge of views in the domain. Since domain is in the center and the ambition to support multiple views per domain.

eliykat reviewed (commented):
Looks good to me, but I'd like a final review from @MGibson1.

MGibson1 commented on apps/browser/src/popup/settings/folder-add-edit.component.ts:
```diff
@@ -22,9 +23,17 @@ export class FolderAddEditComponent extends BaseFolderAddEditComponent {
+      i18nService,
+      platformUtilsService,
+      logService,
+      cryptoService
```
I think this review comment got buried, but I don't think components should use CryptoService. `FolderApiService` should instead accept a `FolderView` and encrypt it prior to upload.

eliykat commented:
Removing myself from reviewers on the understanding this will be done in the SDK instead.
